src/EditorFeatures/Core/Implementation/IntelliSense/QuickInfo/IntellisenseQuickInfoBuilder.cs

```diff
@@ -47,18 +55,90 @@ internal static IntellisenseQuickInfoItem BuildItem(ITrackingSpan trackingSpan,
                 quickInfoItem.Sections.Where(s => s.Kind != QuickInfoSectionKinds.Description)
                                       .Select(BuildClassifiedTextElement));
 
+            // build text for RelatedSpan
+            if (quickInfoItem.RelatedSpans.Any())
+            {
+                elements.Add(await BuildTextElementForRelatedSpans(quickInfoItem, snapshot, document, cancellationToken).ConfigureAwait(false));
+            }
+
             var content = new ContainerElement(
                                 ContainerElementStyle.Stacked,
                                 elements);
 
             return new IntellisenseQuickInfoItem(trackingSpan, content);
         }
 
+        private static async Task<ClassifiedTextElement> BuildTextElementForRelatedSpans(CodeAnalysisQuickInfoItem quickInfoItem, ITextSnapshot snapshot, Document document, CancellationToken cancellationToken)
+        {
+            var classificationService = document.Project.LanguageServices.GetService<IClassificationService>();
+
+            var classifiedSpans = new List<ClassifiedSpan>();
```

Comment:
i'm really curious why all this is needed.  what caused this to stop working?

Reply:
we are moving to new editor QuickInfo API

Reply:
So the new quickinfo APi doesn't allow you to just supply a WPF components  (like a IWpfTextView?)?

Reply:
I believe they are moving to data API. returning UI make it to have hard dependency on UI stack and not pure data API (marshaling between process, wire, app domain not possible)

Reply:
using project buffer + view is good and lazy, but it also only works if host has setup Roslyn right like registering buffer for the view to certain workspace and the workspace having solution to provide all semantic info and etc.

basically, it require something else to present. this way, everything is already processed and host can show this even without Roslyn being there. sure, drawback is it is not as rich as showing real buffer (and ever tagger like squiggles and etc)

Reply:
Note: my ideal would be that a language coudl still describe that it wants to see a sub-section of a view.  Then, the actual component in the host cna decide waht to do.  If the host supports presenting a 'view', it could just show that.  If the host does *not* support it, then we can have the logic to build up a string as appropriate.   :)

> basically, it require something else to present. this way, everything is already processed and host can show this even without Roslyn being there. sure, drawback is it is not as rich as showing real buffer (and ever tagger like squiggles and etc)


Right.  it seems like we're going with "lowest common denominator".  My preference would be to have the rich info there, perhaps as something the host can ask for...

Otherwise, it feels like we're regressing, when we could just support both scenarios.  Thoughts?

Reply:
since it is editor api which doesn't concern about any other file than the buffer open, to support such thing,  VS will need concept of snapshot of whole solution which it doesn't have right now.

Reply:
Sorry... i'm not understanding what you mean... 
Can you clarify?

Reply:
Yeah, i think i'm missing something.  Why can't the editor also supply an API that allows you to present a section of the buffer in quick info?

This is actually a really nice experience.  It's not some 'transformed' set of data . It's the real buffer.

--

I totally get that some hosts won't support this.  And we should have a good story tehre.  But it seems unfortunate that we're actively degrading the VS scenario when thsi was so nice before.  Why don't we just do the work to make that scenario work again?

Reply:
Technically, the only thing that's regressed here is that the new quick info API is called on a background thread. If you are tolerant of a bit of hackery, there is nothing that stops you from marshalling to the UI thread and returning WPF, but I'd strongly discourage that. The 'right way' is to define a custom transformations from your custom data type into custom UI. See the [API wiki for info](https://github.com/Microsoft/vs-editor-api/wiki/Modern-ToolTip-API#supported-conversions). NOTE: this would require a complimentary change to be made for VS for Mac and VS Live Share.

>It seems unfortunate that we cannot have 'pure data API' that represents the part of the view to show. But, i can see why they may not want that.

This seems like a perfectly reasonable ask, the only reason it wasn't part of the original design is that we already support representations of classified text, icons, basic layouts, and the bag of supported formats was getting uncomfortably large. Before, defining your custom data -> UI transformation was the recommendation, but now I'd encourage just using the 3 build in supported types so that the data you provide is easily serializable.

Reply:
custom type/transformation seems good way to do it. since then it becomes implementation detail and quick info provider choose whatever they want from it rather than contract is built into quick info API.

basically data api return custom data which has info on necessary snapshot and span and whathever and it also provide custom UI factory that knows how to interpret its custom data and create UI for it. and it is custom for each hosts.

Reply:
Yep, that's one way. The only thing is that that rather elegant solution fell apart when VS Live Share entered the scene. We're still working on the right way to handle serializing arbitrary pieces of data. The story might end up being that you conform to a small set of known types or your content is dropped.

Reply:
ya. I am not sure VS live or vscode case whether this projection buffer + view will ever work since that assumes that there is roslyn workspace that tracks which buffer map to which roslyn solution. without that, view will only show black texts.

and I don't think VS live or vscode ever will have that info in the host. this will just work for VS as it is now. but I think good to know if we want, we can support VS host specific experience.

Reply:
Yep, VS and Windows specific experiences are still supported as before but I'd highly encourage moving away from them. It may require some hard decisions.